### ISSUE TITLE
Edit a single occurence of a recurrent event

### ISSUE THREAD

**Issue by julien4215** [python-caldav/caldav#379]
Is it possible to edit a single occurrence of a recurrent event ? I couldn't find a solution.

I guess this is linked to the issue #35.

**Comment by tobixen** [comment]
No, editing a single instance is more trivial than deleting a single instance.  As for now there is no explicit way of doing it in the caldav library (arguably it's outside the scope of the caldav library since this is more in the icalendar layer than the caldav layer - but usability beats purity, so I would be willing to accept a pull request on that if it's done in a good way).

I think the best way to do this is to use `calendar.search(begin=..., end=..., expand=True)` to find the recurrence instance you want to edit - then when you got it, you may edit it and save it back to the calendar.  The server should take care that only that instance is affected.

**Comment by julien4215** [comment]
I tried to do this with nextcloud calendar but when I save the single occurence I edited, it deletes the recurrence event and keeps only the single ocurrence.

**Comment by tobixen** [comment]
I should write up some test code for this and see how different servers behave.  Unfortunately I can't promise that I will have time for it anytime soon.

**Comment by julien4215** [comment]
Sure, I understand. Thank you for your quick answer.

**Comment by julien4215** [comment]
Thanks a lot for the explanations ! I will try this solution.

**Comment by julien4215** [comment]
It worked ! 
Just to clarify if someone faces the same issue, there is a typo in the property, it is `recurrence-id` and it shouldn't be missing on the recurrence instance (events obtained by expanding the recurrence) but there is no `recurrence-id` property for the recurrence event.

**Comment by tobixen** [comment]
Right, I made a mistake in the post above, of course there should be no underscore in `RECURRENCE-ID` (capital letters according to the standard, but the icalendar library is case insensitive)

**Comment by tobixen** [comment]
Yesterday I got a personal email on this.  So to reiterate, it should be possible to do something like this (completely untested, so probably lots of mistakes there) to edit a single instance:

```
my_search_results = calendar.search(begin=..., end=..., uid=...., expand=True)
assert(len(my_search_results)) == 1
event = my_search_results[0]
event.icalendar_component['summary'] = "Monthly meeting - this time with restaurant visit"
event.icalendar_component['location'] = "The Restaurant at the corner"
event.save()
```

I'm too lazy to check if the RFC states anything about this, but most likely the code above won't do what it should do on most calendar servers/providers.  However, with some extra code it should most likely work correctly on (almost) all calendar servers/providers, something like this (I'm too lazy to check if the code makes sense and even look up what methods to use):

```python
if 'RECURRENCE-ID' in self.icalendar_component:
    ## Find the recurrent object, as it is stored on the server
    ## (we get the base definition + list of exceptions back)
    canonical = self.parent.object_by_uid(self.icalendar_component['UID'])

    ## Remove the instance we're editing, if it already exists:
    icalobj = canonical.icalendar_instance
    for x in icalobj.subcomponents:
        if x['RECURRENCE-ID'] == self.icalendar_component['RECURRENCE_ID']:
           icalobj.remove(x)
    
    ## Add the instance to be saved:
   icalobj.add_subcomponent(self.icalendar_component)
   
   ## Send everything to the server
   canonical.save()
   return
```
